File: app/src/main/java/com/zuk0/gaijinsmash/riderz/data/remote/repository/StationRepository.java
package com.zuk0.gaijinsmash.riderz.data.remote.repository;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import android.util.Log;

import com.zuk0.gaijinsmash.riderz.data.local.room.dao.StationDao;
import com.zuk0.gaijinsmash.riderz.data.local.entity.station_response.Station;
import com.zuk0.gaijinsmash.riderz.data.local.entity.station_response.StationXmlResponse;
import com.zuk0.gaijinsmash.riderz.data.remote.retrofit.BartService;

import java.io.IOException;
import java.util.concurrent.Executor;

import javax.inject.Inject;
import javax.inject.Named;
import javax.inject.Singleton;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

@Singleton
public class StationRepository  {

    private final BartService service;
    private final StationDao stationDAO;
    private final Executor executor;

    @Inject
    StationRepository(BartService service, StationDao stationDAO, Executor executor) {
        this.service = service;
        this.stationDAO = stationDAO;
        this.executor = executor;
    }

    public LiveData<StationXmlResponse> getStation(String abbr) {
        //todo: if cached != null, return cached

        final MutableLiveData<StationXmlResponse> data = new MutableLiveData<>();
        service.getStation(abbr).enqueue(new Callback<StationXmlResponse>() {
            @Override
            public void onResponse(@NonNull Call<StationXmlResponse> call, @NonNull Response<StationXmlResponse> response) {
                data.postValue(response.body());
                Log.i("Station", response.message());
            }

            @Override
            public void onFailure(@NonNull Call<StationXmlResponse> call, @NonNull Throwable t) {
                Log.wtf("onFailure", "station: " + t.getMessage());
            }
        });
        return data;
    }

    public LiveData<StationXmlResponse> getStations() {
        final MutableLiveData<StationXmlResponse> data = new MutableLiveData<>();
        service.getAllStations().enqueue(new Callback<StationXmlResponse>() {
            @Override
            public void onResponse(@NonNull Call<StationXmlResponse> call, @NonNull Response<StationXmlResponse> response) {
                data.postValue(response.body());
                Log.i("Stations", response.message());
            }

            @Override
            public void onFailure(@NonNull Call<StationXmlResponse> call, @NonNull Throwable t) {
                Log.wtf("StationRepository", t.getMessage());
            }
        });
        return data;
    }

    public void refreshStation(final String name) {
        executor.execute(() -> {
            //todo: check if station exists, then check if station data exists
          boolean stationExists = stationDAO.getStationByName(name) != null;
          String stationData = stationDAO.getStationByName(name).getIntro();
          boolean stationDataExists = !stationData.equals("");
          if(!stationExists) {
              // refresh data
              try {
                  Response response = service.getStation(name).execute();
                  stationDAO.save((Station) response.body());
              } catch (IOException e) {
                  e.printStackTrace();
              }
          }
        });
    }
}

File: app/src/main/java/com/zuk0/gaijinsmash/riderz/data/remote/repository/StationRepository.kt
package com.zuk0.gaijinsmash.riderz.data.remote.repository

import android.util.Log
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.zuk0.gaijinsmash.riderz.data.local.entity.station_response.Station
import com.zuk0.gaijinsmash.riderz.data.local.entity.station_response.StationXmlResponse
import com.zuk0.gaijinsmash.riderz.data.local.room.dao.StationDao
import com.zuk0.gaijinsmash.riderz.data.remote.retrofit.BartService
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.io.IOException
import java.util.concurrent.Executor
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class StationRepository
@Inject constructor(val service: BartService, val stationDAO: StationDao, val executor: Executor) {
    fun getStation(abbr: String): LiveData<StationXmlResponse> { //todo: if cached != null, return cached
        val data = MutableLiveData<StationXmlResponse>()
        service?.getStation(abbr)?.enqueue(object : Callback<StationXmlResponse> {
            override fun onResponse(call: Call<StationXmlResponse>, response: Response<StationXmlResponse>) {
                data.postValue(response.body())
                Log.i("Station", response.message())
            }

            override fun onFailure(call: Call<StationXmlResponse>, t: Throwable) {
                Log.wtf("onFailure", "station: " + t.message)
            }
        })
        return data
    }

    val stations: LiveData<StationXmlResponse>
        get() {
            val data = MutableLiveData<StationXmlResponse>()

            service?.getAllStations()?.enqueue(object : Callback<StationXmlResponse> {
                override fun onResponse(call: Call<StationXmlResponse>, response: Response<StationXmlResponse>) {
                    data.postValue(response.body())
                    Log.i("Stations", response.message())
                }

                override fun onFailure(call: Call<StationXmlResponse>, t: Throwable) {
                    Log.wtf("StationRepository", t.message)
                }
            })
            return data
        }

    fun refreshStation(name: String?) {
      /*  executor.execute {
            //todo: check if station exists, then check if station data exists
            val stationExists = stationDAO.getStationByName(name) != null
            val stationData = stationDAO.getStationByName(name)!!.intro
            val stationDataExists = stationData != ""
            if (!stationExists) { // refresh data
                try {
                    val response: Response<*> = service.getStation(name).execute()
                    stationDAO.save(response.body() as Station?)
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            }
        }*/
    }

}